File: backend/app/schemas/templates.py
"""
Template Schemas for SecureVault

Handles project template definitions, folder structures, and template management.
"""

from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any, Union
from datetime import datetime
from enum import Enum


class TemplateCategory(str, Enum):
    """Template categories for organization"""
    BUSINESS = "business"
    LEGAL = "legal"
    DEVELOPMENT = "development"
    MARKETING = "marketing"
    HR = "hr"
    FINANCE = "finance"
    RESEARCH = "research"
    EVENT = "event"
    GENERAL = "general"


class FolderStructureItem(BaseModel):
    """Individual folder or document in template structure"""
    name: str = Field(..., description="Name of the folder/document")
    type: str = Field(..., description="'folder' or 'document'")
    description: Optional[str] = Field(None, description="Description of this item")
    tags: List[str] = Field(default_factory=list, description="Default tags for this item")
    children: List['FolderStructureItem'] = Field(default_factory=list, description="Nested items")
    document_template: Optional[str] = Field(None, description="Template content for documents")
    permissions: Optional[Dict[str, Any]] = Field(None, description="Default permissions")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="Additional metadata")
    
    class Config:
        json_schema_extra = {
            "example": {
                "name": "Requirements",
                "type": "folder",
                "description": "Project requirements and specifications",
                "tags": ["requirements", "planning"],
                "children": [
                    {
                        "name": "Business Requirements.md",
                        "type": "document",
                        "description": "Business requirements document",
                        "tags": ["requirements", "business"],
                        "document_template": "# Business Requirements\n\n## Overview\n\n## Stakeholders\n\n## Requirements"
                    }
                ]
            }
        }

# Fix forward reference
FolderStructureItem.model_rebuild()


class TemplateVariables(BaseModel):
    """Variables that can be used in template names and content"""
    project_name: Optional[str] = Field(None, description="Project/client name")
    date: Optional[str] = Field(None, description="Current date")
    year: Optional[str] = Field(None, description="Current year")
    month: Optional[str] = Field(None, description="Current month")
    department: Optional[str] = Field(None, description="Department name")
    custom_vars: Dict[str, str] = Field(default_factory=dict, description="Custom variables")


class ProjectTemplate(BaseModel):
    """Project template definition"""
    id: Optional[str] = Field(None, description="Template ID")
    name: str = Field(..., description="Template name")
    description: str = Field(..., description="Template description")
    category: TemplateCategory = Field(..., description="Template category")
    icon: Optional[str] = Field(None, description="Icon name for UI")
    color: Optional[str] = Field(None, description="Color theme for UI")
    
    # Template structure
    structure: List[FolderStructureItem] = Field(..., description="Folder structure definition")
    
    # Default settings
    default_tags: List[str] = Field(default_factory=list, description="Default tags applied to root folder")
    default_permissions: Optional[Dict[str, Any]] = Field(None, description="Default permissions")
    default_metadata: Dict[str, Any] = Field(default_factory=dict, description="Default metadata")
    
    # Template configuration
    variables: List[str] = Field(default_factory=list, description="Available variables for this template")
    allow_customization: bool = Field(True, description="Allow users to modify structure")
    is_system_template: bool = Field(False, description="System-provided template (non-editable)")
    
    # Audit fields
    created_by: Optional[int] = Field(None, description="User ID who created template")
    created_at: Optional[datetime] = Field(None, description="Creation timestamp")
    updated_by: Optional[int] = Field(None, description="User ID who last updated")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")
    version: int = Field(1, description="Template version")
    
    # Usage statistics
    usage_count: int = Field(0, description="Number of times template has been used")
    last_used: Optional[datetime] = Field(None, description="Last time template was used")

    class Config:
        json_schema_extra = {
            "example": {
                "name": "Software Development Project",
                "description": "Complete folder structure for software development projects",
                "category": "development",
                "icon": "code",
                "color": "blue",
                "structure": [
                    {
                        "name": "Requirements",
                        "type": "folder",
                        "description": "Project requirements and specifications",
                        "tags": ["requirements", "planning"],
                        "children": [
                            {
                                "name": "Business Requirements.md",
                                "type": "document",
                                "tags": ["requirements", "business"]
                            }
                        ]
                    },
                    {
                        "name": "Design",
                        "type": "folder",
                        "description": "System design and architecture",
                        "tags": ["design", "architecture"]
                    }
                ],
                "default_tags": ["project", "development"],
                "variables": ["project_name", "date", "department"]
            }
        }


class TemplateCreateRequest(BaseModel):
    """Request to create a new template"""
    name: str = Field(..., description="Template name")
    description: str = Field(..., description="Template description")
    category: TemplateCategory = Field(..., description="Template category")
    icon: Optional[str] = Field(None, description="Icon name")
    color: Optional[str] = Field(None, description="Color theme")
    structure: List[FolderStructureItem] = Field(..., description="Folder structure")
    default_tags: List[str] = Field(default_factory=list)
    default_permissions: Optional[Dict[str, Any]] = Field(None)
    default_metadata: Dict[str, Any] = Field(default_factory=dict)
    variables: List[str] = Field(default_factory=list)
    allow_customization: bool = Field(True)


class TemplateUpdateRequest(BaseModel):
    """Request to update an existing template"""
    name: Optional[str] = Field(None, description="Template name")
    description: Optional[str] = Field(None, description="Template description")
    category: Optional[TemplateCategory] = Field(None, description="Template category")
    icon: Optional[str] = Field(None, description="Icon name")
    color: Optional[str] = Field(None, description="Color theme")
    structure: Optional[List[FolderStructureItem]] = Field(None, description="Folder structure")
    default_tags: Optional[List[str]] = Field(None)
    default_permissions: Optional[Dict[str, Any]] = Field(None)
    default_metadata: Optional[Dict[str, Any]] = Field(None)
    variables: Optional[List[str]] = Field(None)
    allow_customization: Optional[bool] = Field(None)


class TemplateApplicationRequest(BaseModel):
    """Request to apply a template to create folder structure"""
    template_id: str = Field(..., description="Template ID to apply")
    parent_folder_id: Optional[int] = Field(None, description="Parent folder ID (null for root)")
    target_folder_id: Optional[int] = Field(None, description="Apply template to existing folder (alternative to creating new folder)")
    variables: TemplateVariables = Field(default_factory=TemplateVariables, description="Variable values")
    custom_name: Optional[str] = Field(None, description="Custom name for root folder (ignored if target_folder_id is provided)")
    apply_permissions: bool = Field(True, description="Apply template permissions")
    apply_tags: bool = Field(True, description="Apply template tags")
    create_documents: bool = Field(False, description="Create placeholder documents")


class TemplateApplicationResult(BaseModel):
    """Result of template application"""
    success: bool = Field(..., description="Whether application succeeded")
    root_folder_id: Optional[int] = Field(None, description="ID of created root folder")
    created_folders: List[int] = Field(default_factory=list, description="IDs of all created folders")
    created_documents: List[int] = Field(default_factory=list, description="IDs of any created documents")
    errors: List[str] = Field(default_factory=list, description="Any errors encountered")
    warnings: List[str] = Field(default_factory=list, description="Any warnings")


class TemplateListResponse(BaseModel):
    """Response for template listing"""
    templates: List[ProjectTemplate] = Field(..., description="List of templates")
    total: int = Field(..., description="Total number of templates")
    categories: List[str] = Field(..., description="Available categories")


class TemplatePreview(BaseModel):
    """Preview of what a template will create"""
    template: ProjectTemplate = Field(..., description="Template being previewed")
    preview_structure: List[Dict[str, Any]] = Field(..., description="Preview of folder structure with resolved variables")
    estimated_folders: int = Field(..., description="Number of folders that will be created")
    estimated_documents: int = Field(..., description="Number of documents that will be created")
    resolved_variables: Dict[str, str] = Field(..., description="Variable values that will be used")

File: frontend/src/services/api/templates.ts
/**
 * Template API Service for SecureVault
 * 
 * Provides API methods for project template management including:
 * - Template listing and filtering
 * - Template preview and application
 * - Custom template management
 */

import { apiRequest } from '../api';

export interface TemplateCategory {
  BUSINESS: 'business';
  LEGAL: 'legal';
  DEVELOPMENT: 'development';
  MARKETING: 'marketing';
  HR: 'hr';
  FINANCE: 'finance';
  RESEARCH: 'research';
  EVENT: 'event';
  GENERAL: 'general';
}

export interface FolderStructureItem {
  name: string;
  type: 'folder' | 'document';
  description?: string;
  tags: string[];
  children: FolderStructureItem[];
  document_template?: string;
  permissions?: Record<string, any>;
  metadata: Record<string, any>;
}

export interface ProjectTemplate {
  id?: string;
  name: string;
  description: string;
  category: keyof TemplateCategory;
  icon?: string;
  color?: string;
  structure: FolderStructureItem[];
  default_tags: string[];
  default_permissions?: Record<string, any>;
  default_metadata: Record<string, any>;
  variables: string[];
  allow_customization: boolean;
  is_system_template: boolean;
  created_by?: number;
  created_at?: string;
  updated_by?: number;
  updated_at?: string;
  version: number;
  usage_count: number;
  last_used?: string;
}

export interface TemplateVariables {
  project_name?: string;
  date?: string;
  year?: string;
  month?: string;
  department?: string;
  custom_vars: Record<string, string>;
}

export interface TemplateApplicationRequest {
  template_id: string;
  parent_folder_id?: number | null;
  target_folder_id?: number | null; // Apply template to existing folder
  variables: TemplateVariables;
  custom_name?: string; // For creating new folder (ignored if target_folder_id is provided)
  apply_permissions: boolean;
  apply_tags: boolean;
  create_documents: boolean;
}

export interface TemplateApplicationResult {
  success: boolean;
  root_folder_id?: number;
  created_folders: number[];
  created_documents: number[];
  errors: string[];
  warnings: string[];
}

export interface TemplateListResponse {
  templates: ProjectTemplate[];
  total: number;
  categories: string[];
}

export interface TemplatePreview {
  template: ProjectTemplate;
  preview_structure: any[];
  estimated_folders: number;
  estimated_documents: number;
  resolved_variables: Record<string, string>;
}

export class TemplatesApiService {
  /**
   * List all available templates
   */
  async listTemplates(category?: string, includeCustom: boolean = true): Promise<TemplateListResponse> {
    const params = new URLSearchParams();
    if (category) {
      params.append('category', category);
    }
    params.append('include_custom', includeCustom.toString());

    const response = await apiRequest<TemplateListResponse>('GET', `/api/v1/templates/?${params}`);
    
    if (!response.success) {
      throw new Error(response.error?.detail || 'Failed to load templates');
    }
    
    return response.data!;
  }

  /**
   * Get a specific template by ID
   */
  async getTemplate(templateId: string): Promise<ProjectTemplate> {
    const response = await apiRequest<ProjectTemplate>('GET', `/api/v1/templates/${templateId}`);
    
    if (!response.success) {
      throw new Error(response.error?.detail || 'Failed to get template');
    }
    
    return response.data!;
  }

  /**
   * Preview what a template will create
   */
  async previewTemplate(templateId: string, variables: TemplateVariables): Promise<TemplatePreview> {
    const response = await apiRequest<TemplatePreview>('POST', `/api/v1/templates/${templateId}/preview`, variables);
    
    if (!response.success) {
      throw new Error(response.error?.detail || 'Failed to preview template');
    }
    
    return response.data!;
  }

  /**
   * Apply a template to create folder structure
   */
  async applyTemplate(request: TemplateApplicationRequest): Promise<TemplateApplicationResult> {
    const response = await apiRequest<TemplateApplicationResult>(
      'POST', 
      `/api/v1/templates/${request.template_id}/apply`, 
      request
    );
    
    if (!response.success) {
      throw new Error(response.error?.detail || 'Failed to apply template');
    }
    
    return response.data!;
  }

  /**
   * Get all template categories
   */
  async getTemplateCategories(): Promise<string[]> {
    const response = await apiRequest<string[]>('GET', '/api/v1/templates/categories/');
    
    if (!response.success) {
      throw new Error(response.error?.detail || 'Failed to get categories');
    }
    
    return response.data!;
  }

  /**
   * Create a new custom template
   */
  async createTemplate(templateData: Omit<ProjectTemplate, 'id' | 'created_by' | 'created_at' | 'updated_by' | 'updated_at' | 'version' | 'usage_count' | 'last_used'>): Promise<ProjectTemplate> {
    const response = await apiRequest<ProjectTemplate>('POST', '/api/v1/templates/', templateData);
    
    if (!response.success) {
      throw new Error(response.error?.detail || 'Failed to create template');
    }
    
    return response.data!;
  }

  /**
   * Update an existing template
   */
  async updateTemplate(templateId: string, templateData: Partial<ProjectTemplate>): Promise<ProjectTemplate> {
    const response = await apiRequest<ProjectTemplate>('PUT', `/api/v1/templates/${templateId}`, templateData);
    
    if (!response.success) {
      throw new Error(response.error?.detail || 'Failed to update template');
    }
    
    return response.data!;
  }

  /**
   * Delete a template
   */
  async deleteTemplate(templateId: string): Promise<void> {
    const response = await apiRequest<void>('DELETE', `/api/v1/templates/${templateId}`);
    
    if (!response.success) {
      throw new Error(response.error?.detail || 'Failed to delete template');
    }
  }

  /**
   * Get template usage statistics
   */
  async getTemplateUsageStats(templateId: string): Promise<{
    template_id: string;
    template_name: string;
    usage_count: number;
    last_used?: string;
    created_at?: string;
    version: number;
  }> {
    const response = await apiRequest<any>('GET', `/api/v1/templates/${templateId}/usage-stats`);
    
    if (!response.success) {
      throw new Error(response.error?.detail || 'Failed to get template statistics');
    }
    
    return response.data!;
  }
}

export const templatesApi = new TemplatesApiService();